### PR TITLE
Remove consecutive functional inverses to optimize

### PR DESCRIPTION
**Pull request by clorenz7**
See:
https://github.com/Theano/Theano/issues/864

Was curious about best practices for inverse pairs like Log and Exp, because Log will return a NaN if the input is less than 0.

### CODE REVIEW

**Comment by nouiz** [comment]
There is some errors by travis. Tests should pass when the theano flags mode=FAST_COMPILE is set. It is related to that case.

File: theano/tensor/tests/test_opt.py [hunk 1]
```diff
@@ -3812,6 +3812,56 @@ def test(self):
...
+    def setUp(self):
+        mode = theano.compile.get_default_mode()
+        self.mode = mode.including('local_func_inv')
+
+    def test(self):
+        """
+        test that consecutive ops that are functional inverses are removed
+        """
+
+        x = T.fmatrix()
```
Comment (nouiz) [inline theano/tensor/tests/test_opt.py]: You need to also tests the other inverse functions that you optimize.

File: theano/tensor/tests/test_opt.py [hunk 2]
```diff
@@ -3812,6 +3812,56 @@ def test(self):
...
+
+        # Test that non-inverse functions are ran normally
+        x = T.fmatrix()
+        o = T.rad2deg(T.rad2deg(x))
+        f = theano.function([x], o, mode=self.mode)
+        dx = numpy.random.rand(5, 4).astype("float32")+0.01
+        delta = f(dx) - dx
+        topo = f.maker.fgraph.toposort()
+
+        self.assertEqual(len(topo), 1)
```
Comment (nouiz) [inline theano/tensor/tests/test_opt.py]: The travis log are here: https://s3.amazonaws.com/archive.travis-ci.org/jobs/58345808/log.txt

The failing test:
```
======================================================================
FAIL: test that consecutive ops that are functional inverses are removed
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/tests/test_opt.py", line 3859, in test
    self.assertEqual(len(topo), 1)
AssertionError: 2 != 1
----------------------------------------------------------------------
```

In FAST_COMPILE, the topo have a lenght of 2, so you could just relax this assert to test for the len of topo to be 1 or 2.

Thanks for the PR.

**Comment by clorenz7** [comment]
@nouiz The unit tests should pass now, I wasn't using theano-nose locally.

**Comment by nouiz** [comment]
thanks!

**Comment by nouiz** [comment]
@lamblin what do you think of the question in the comment of the PR? Personnaly, I would not do them until someone really need them, as I do not see a good answer.

Some talk with @abergeron came to: remove it and add an assert op.

**Comment by lamblin** [comment]
I agree with removing it and adding an `assert`.

**Pull request merged**